common/os_calls.c

```diff
@@ -3014,6 +3033,21 @@ g_set_allusercontext(int uid)
     return (rv != 0);  /* Return 0 or 1 */
 }
 #endif
+
+/*****************************************************************************/
+void
+g_get_executable_path(char *buf, int bufsize)
+{
+#if defined(__APPLE__)
+    uint32_t _bufsize = bufsize;
+    _NSGetExecutablePath(buf, &_bufsize);
+#else
+    LOG(LOG_LEVEL_WARN, "g_get_executable_path(): not implemented yet!");
+    buf = strdup("xrdp");
+#endif
```

Comment:
This won't work. The result of the `strdup()` is called in the local variable buf, which is lost on return. Not only do you not get a name, there's a memory leak.

Can we add an extra parameter to the front of this, to make the whole thing more portable?

In os_calls.h:-

```c
enum xrdp_exe
{
    E_XE_XRDP = 0
    // Add others below
}
```
Then :-

```c
void
g_get_executable_path(enum xrdp_exe xe, char *buf, int bufsize)
{
    switch(xe)
    {
        case E_XE_XRDP:
            snprintf(buf, bufsize, XRDP_SBIN_PATH "/xrdp");
        default:
            :PG(LOG_LEVEL_WARN, "g_get_executable_path(): Unsupported exe %d", (int)xe);
            if (buf_size > 0)
            {
                buf[0] = '\0';
            }
    }
}
```

This is more generic, and lets us use if for (e.g.) sesman calling chansrv.